AX: scroll views shouldn't be ignored when they have a remote frame child

hoffmanjoshua opened the pull request:
#### 03436a4ea53a9e093c3b626fb60fac615edf40c4
<pre>
AX: scroll views shouldn&apos;t be ignored when they have a remote frame child
<a href="https://bugs.webkit.org/show_bug.cgi?id=284379">https://bugs.webkit.org/show_bug.cgi?id=284379</a>
<a href="https://rdar.apple.com/problem/141228277">rdar://problem/141228277</a>

Reviewed by Tyler Wilcock.

We can&apos;t ignore scroll view&apos;s when they have remote children—this leads to missing content
on the isolated tree, since this scroll view is being requested.

A new testing method was added to AccessibilityUIElement, childAtIndexWithRemoteElement,
for validating the existence of a remote frame.

* LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html:
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-returns-remote-frame.html:
* LayoutTests/http/tests/site-isolation/accessibility/remote-element-returned-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/accessibility/remote-element-returned.html: Added.
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::computeIsIgnored const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(makeNSArray):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityArrayAttributeValues:index:maxCount:]):
(-[WebAccessibilityObjectWrapper _accessibilityChildrenFromIndex:maxCount:returnPlatformElements:]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrame):
(WTR::AccessibilityUIElement::childAtIndexWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting): Deleted.
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrameForTesting): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::childAtIndexWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrame):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting): Deleted.
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrameForTesting): Deleted.

Canonical link: <a href="https://commits.webkit.org/287701@main">https://commits.webkit.org/287701@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92e5a446847ff09c59cf7d93d9108af06316050c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80524 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59531 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85045 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68592 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7836 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62912 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20718 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83593 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73310 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43215 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50322 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27468 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29965 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71467 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27991 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86479 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71204 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7925 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69147 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70444 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17559 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14458 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13403 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7712 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13231 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7551 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11070 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9356 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/http/tests/site-isolation/accessibility/remote-element-returned.html:
```diff
@@ -0,0 +1,42 @@
+            }
+        }, 0);
+    });
+}
```
Is it possible for us to include `LayoutTests/resources/accessibility-helpers.js` rather than re-defining this function?

hoffmanjoshua replied:
Updated! (for http tests, `js-test-resources` is a required alias for accessing `LayoutTest/resources`)

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -214,6 +214,9 @@ void AccessibilityScrollView::clearChildren()
 
 bool AccessibilityScrollView::computeIsIgnored() const
 {
+    if (m_remoteFrame)
```
Could we please add a comment explaining why having a remote frame means we shouldn't ignore this scroll view?

twilco commented on Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
```diff
@@ -94,6 +94,7 @@ class AccessibilityUIElement : public JSWrappable {
 
     JSValueRef children(JSContextRef);
     RefPtr<AccessibilityUIElement> childAtIndex(unsigned);
+    RefPtr<AccessibilityUIElement> childAtIndexWithRemoteElementForTesting(unsigned);
```
Is the `ForTesting` suffix necessary here? I feel like everything inside `AccessibilityUIElement` is implicitly for testing, because it cannot be used in any other way 🤔

hoffmanjoshua replied:
The `forTesting` suffix I have been using to distinguish behaviors that the test exhibits that actual platforms wouldn't experience (for example, we will always return platform elements in non-testing scenarios). But if we think it's redundant, I can definitely remove it!

twilco replied:
Ahh, gotcha. Well I'd be fine with either removing the suffix, or adding a comment explaining the name (as you've done here) 👍

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -4013,60 +4013,69 @@ - (pid_t)accessibilityPresenterProcessIdentifier
+- (NSArray *)_accessibilityChildrenFromIndex:(NSUInteger)index maxCount:(NSUInteger)maxCount returnPlatformElements:(BOOL)returnPlatformElements
+{
+    if (!returnPlatformElements && !AXObjectCache::clientIsInTestMode())
+        return nil;
```
Do we need this guard? Seems like ATs would only hit this if they called this private method, which legitimate ATs would never do. So then the question becomes, could a misbehaving AT be able to do something bad if they call this private method with `returnPlatformElements == NO`?

hoffmanjoshua replied:
Probably not—this was just to be extra safe. But since this method is private, we can probably remove it.

hoffmanjoshua committed:
AX: scroll views shouldn't be ignored when they have a remote frame child
https://bugs.webkit.org/show_bug.cgi?id=284379
rdar://problem/141228277

Reviewed by Tyler Wilcock.

We can't ignore scroll view's when they have remote children—this leads to missing content
on the isolated tree, since this scroll view is being requested.

A new testing method was added to AccessibilityUIElement, childAtIndexWithRemoteElement,
for validating the existence of a remote frame.

* LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html:
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-returns-remote-frame.html:
* LayoutTests/http/tests/site-isolation/accessibility/remote-element-returned-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/accessibility/remote-element-returned.html: Added.
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::computeIsIgnored const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(makeNSArray):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityArrayAttributeValues:index:maxCount:]):
(-[WebAccessibilityObjectWrapper _accessibilityChildrenFromIndex:maxCount:returnPlatformElements:]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrame):
(WTR::AccessibilityUIElement::childAtIndexWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting): Deleted.
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrameForTesting): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::childAtIndexWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElement):
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrame):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting): Deleted.
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrameForTesting): Deleted.

Canonical link: https://commits.webkit.org/287701@main
